File: src/main/java/com/jxkj/base/base/ui/activity/AbstractSimpleActivity.java
package com.jxkj.base.base.ui.activity;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.gyf.barlibrary.ImmersionBar;
import com.jxkj.base.frame.R;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import me.yokeyword.fragmentation.SupportActivity;

public abstract class AbstractSimpleActivity extends SupportActivity implements IActivity {
    private Unbinder unBinder;
    protected ImmersionBar immersionBar;
    protected TextView tvTitle;
    protected boolean isTransparent = true;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
        View layBack = findViewById(R.id.lay_back);
        if (layBack != null) {
            layBack.setOnClickListener((view) -> this.finish());
        }
        unBinder = ButterKnife.bind(this);
        initStatusBar();
        onViewCreated();
        initToolbar();
        initView();
        initEventAndData();
    }

    protected void setImmersionBarDark(boolean isDark) {
        immersionBar.navigationBarDarkIcon(isDark).statusBarDarkFont(isDark);
        try {
            Class<?> barClass = immersionBar.getClass();
            Method method = barClass.getDeclaredMethod("updateBarParams");
            method.setAccessible(true);
            method.invoke(immersionBar);
            method = barClass.getDeclaredMethod("setBar");
            method.setAccessible(true);
            method.invoke(immersionBar);
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException ignore) {
        }
    }

    protected void initStatusBar() {
        immersionBar = ImmersionBar.with(this);//.transparentStatusBar();
        int statusBarColor = statusBarColor();
        if (statusBarColor != -1) {
            immersionBar.statusBarColor(statusBarColor);
        }
        immersionBar
                .statusBarDarkFont(isStatusBarDark())
                .navigationBarDarkIcon(isStatusBarDark())
                .statusBarView(findViewById(R.id.status_bar_view))
                .keyboardEnable(isKeyboardEnable());
        immersionBar.init();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (unBinder != null && unBinder != Unbinder.EMPTY) {
            unBinder.unbind();
            unBinder = null;
        }
        if (immersionBar != null) {
            immersionBar.destroy();
            immersionBar = null;
        }
    }
    protected boolean isStatusBarDark() {
        return false;
    }

    protected boolean isKeyboardEnable() {
        return true;
    }


    protected int statusBarColor() {
        return R.color.theme;
    }
    protected abstract void initView();

    /**
     * 在initEventAndData()之前执行
     */
    protected abstract void onViewCreated();

    /**
     * 获取当前Activity的UI布局
     *
     * @return 布局id
     */
    protected abstract int getLayoutId();

    /**
     * 初始化ToolBar
     */
    protected void initToolbar() {
        TextView tvTitle = findViewById(R.id.toolbar_title);
        if (tvTitle == null) {
            return;
        }
        int titleRes = title();
        if (titleRes != -1) {
            tvTitle.setText(getString(titleRes));
        }
    }


    protected @StringRes
    int title() {
        return -1;
    }

    /**
     * 初始化数据
     */
    protected abstract void initEventAndData();
//    /**
//     * 注册监听
//     */
//    protected abstract void setListener();
}

File: library/src/main/java/com/frame/base/inter/ui/activity/AbstractSimpleActivity.kt
package com.frame.base.inter.ui.activity

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.annotation.StringRes
import com.frame.base.R
import com.frame.base.inter.ui.IView
import com.gyf.barlibrary.ImmersionBar
import me.yokeyword.fragmentation.SupportActivity
import java.lang.reflect.InvocationTargetException

abstract class AbstractSimpleActivity : SupportActivity(), IView {
    protected open var immersionBar: ImmersionBar? = null
    protected var tvTitle: TextView? = null
    protected var isTransparent = true
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        preContentView()
        setContentView()
        val layBack = findViewById<View>(R.id.lay_back)
        layBack?.setOnClickListener { view: View? -> finish() }
        initStatusBar()
        onViewCreated()
        initToolbar()
        initView()
        initEventAndData()
    }

    protected open fun setImmersionBarDark(isDark: Boolean) {
        immersionBar!!.navigationBarDarkIcon(isDark).statusBarDarkFont(isDark)
        try {
            val barClass: Class<*> = immersionBar!!.javaClass
            var method = barClass.getDeclaredMethod("updateBarParams")
            method.isAccessible = true
            method.invoke(immersionBar)
            method = barClass.getDeclaredMethod("setBar")
            method.isAccessible = true
            method.invoke(immersionBar)
        } catch (ignore: NoSuchMethodException) {
        } catch (ignore: InvocationTargetException) {
        } catch (ignore: IllegalAccessException) {
        }
    }

    protected open fun preContentView() {}
    abstract fun setContentView()
    protected open fun initStatusBar() {
        immersionBar = ImmersionBar.with(this) //.transparentStatusBar();

        immersionBar?.run {
            val statusBarColor = statusBarColor()
            if (statusBarColor != -1) {
                statusBarColor(statusBarColor)
            }
            statusBarDarkFont(isStatusBarDark())
                .navigationBarDarkIcon(isStatusBarDark())
                .statusBarView(findViewById(R.id.status_bar_view))
                .keyboardEnable(isKeyboardEnable())
            init()
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        if (immersionBar != null) {
            immersionBar!!.destroy()
            immersionBar = null
        }
    }

    protected open fun isStatusBarDark(): Boolean{
        return true
    }
    protected open fun isKeyboardEnable(): Boolean{
        return true
    }

    protected open  fun statusBarColor(): Int {
        return R.color.white
    }

    protected abstract fun initView()

    /**
     * 在initEventAndData()之前执行
     */
    protected abstract fun onViewCreated()
     override fun loginOut() { }
     override fun showError() {
     }
     override fun showNoNetwork() {
     }
    /**
     * 初始化ToolBar
     */
    protected open fun initToolbar() {
        val tvTitle = findViewById<TextView>(R.id.toolbar_title) ?: return
        val titleRes = title()
        var title: String?
        if (titleRes != -1) {
            tvTitle.text = getString(titleRes)
        } else if (titleStr().also { title = it } != null) {
            tvTitle.text = title
        }
    }

    @StringRes
    protected open fun title(): Int {
        return -1
    }

    protected open fun titleStr(): String? {
        return null
    }

    /**
     * 初始化数据
     */
    abstract fun initEventAndData()
}